src/software/geom/spline.cpp

```diff
@@ -0,0 +1,79 @@
+#include "software/geom/spline.h"
+
+Spline::Spline(const std::vector<Point>& points)
+    : knots(points), domain(std::make_pair(0.0, 1.0))
+{
+    initLinearSegments(points);
+}
+
+Spline::Spline(const std::initializer_list<Point>& points)
+    : knots(points), domain(std::make_pair(0.0, 1.0))
+{
+    initLinearSegments(points);
+}
+
+Point Spline::valueAt(double val) const
+{
+    if (val < domain.first || val > domain.second)
+    {
+        throw std::invalid_argument("val is outside of the domain of this spline");
+    }
+
+    if (val == domain.first)
+    {
+        return knots.front();
+    }
+
+    if (val == domain.second)
+    {
+        return knots.back();
+    }
+
+    // assume that the segments are equally spaced in domain
+    int index = (int)(val * knots.size());
+    if (val < segments[index].start || val > segments[index].end)
+    {
+        throw std::runtime_error("Segments are not equally spaced as expected");
+    }
+    double x_val = segments[index].x.valueAt(val);
+    double y_val = segments[index].y.valueAt(val);
+    return Point(x_val, y_val);
+}
+
+size_t Spline::size(void) const
+{
+    return knots.size();
+}
+
+const std::vector<Point> Spline::getKnots(void) const
+{
+    return knots;
+}
+
+const std::pair<double, double> Spline::getDomain(void) const
+{
+    return domain;
+}
+
+void Spline::initLinearSegments(const std::vector<Point>& points)
```

Comment:
Could we make this a util function? I don't think it _really_ belongs as a class method?

Reply:
it modifies this-> segments, where can I move it?

Reply:
Ah I see now. Eventually we should probably have `Spline(std::vector<Points> points, unsigned int order)` to construct a spline of some order, then this function would be equivalent to calling the constructor with `order=1`. Given that we're trying to keep the scope of this limited though, I'm fine with this as long as we file a ticket to remove it and replace it with a generic structure (and note the ticket here).

Reply:
#983